compiler/rustc_ast/src/attr/mod.rs

```diff
@@ -580,3 +580,12 @@ impl NestedMetaItem {
         MetaItem::from_tokens(tokens).map(NestedMetaItem::MetaItem)
     }
 }
+
+/// Return true if the attributes contain `#[doc(hidden)]`
+pub fn is_doc_hidden(attrs: &[ast::Attribute]) -> bool {
+    attrs
+        .iter()
+        .filter(|attr| attr.has_name(sym::doc))
+        .filter_map(ast::Attribute::meta_item_list)
+        .any(|l| list_contains_name(&l, sym::hidden))
+}
```

Comment:
Could we improve performance if we modified the Ast to explicitly track this? We could still have it part of the attribute list, but have a boolean as part of the item or something so that it gets computed during parsing once, and not every time we compute an item when building the visible_parent_map. I think that might win us some of the performance we're losing back (at the cost of memory).

Reply:
@estebank it was a query before and that seemed to hurt performance more than it helped

Reply:
But queries only help when you call them repeatedly, calling it just once you get the overhead of the query system _and_ still have the O(n) behavior, right? Moving that to be something that happens during parsing removes the O(n) behavior entirely for something that gets evaluated for (I'm guessing) every path, unless I'm missing something?

Reply:
That seems possible? This is only a .6% regression on some small benchmarks though.

Reply:
The wall-time regressions for self-bootstrap are more concerning to me. It's not clear whether those come from newly introduced code, but in combination with the [cycle regressions](https://perf.rust-lang.org/compare.html?calcNewSig=true&start=91d8da1f4ba24679e92b7939a26c681a5d2d3548&end=6a304b8fb7cdacaa12996b2b46a8f61e8a2f1ba7&stat=cycles%3Au) I would not call this "just" .6%.

Do we have any evidence that the regression here is due to doc_hidden being slow? Unfortunately cycles make it a little difficult to do good local comparisons, as cachegrind only really looks at instructions (results locally from an attempt are inconclusive). I can play with some perf-record later, perhaps.

Reply:
Rustc has supressed deprecation warnings for `llvm_asm`. Currently even if warning is supressed, message is calculated.

I think this is where regression for bootstrap time comes.

https://github.com/rust-lang/rust/blob/bb744e1e9f1e193014ab3f04bf5e0640dbe4dc63/compiler/rustc_middle/src/middle/stability.rs#L307

Reply:
That should be easy enough to fix - just add a check to `lint_level_at_node`.